ecosystem/platform/server2/app/models/user.rb

```diff
@@ -0,0 +1,137 @@
+# frozen_string_literal: true
+
+# Copyright (c) Aptos
+# SPDX-License-Identifier: Apache-2.0
+
+class User < ApplicationRecord
+  include RailsStateMachine::Model
+
+  # Include default devise modules. Others available are:
+  # :confirmable, :lockable, :timeoutable, :recoverable,
+  devise :database_authenticatable,
+         :rememberable, :trackable, :validatable,
+         :omniauthable, omniauth_providers: %i[discord github],
+                        authentication_keys: [:username]
+
+  validates :username, uniqueness: { case_sensitive: false }, allow_nil: true
+  validates :email, uniqueness: { case_sensitive: false }, allow_nil: true
```

Comment:
Is it possible to add a constraint that at least one of these is present? Or is it OK for both to be null?

Reply:
it's okay for both to be null: we don't necessarily get a unique thing here from either provider, and so the first thing after connecting an account, will be for them to set a username + email (and then verify it)